File: Cogs/RoleRequestor.py
import discord
from discord.ext import commands

ROLE_REQUEST_CHANNEL = 'role-request'
MODERATOR_ROLE = 'Moderator'


async def is_administrator(ctx: commands.Context):
    return ctx.author.guild_permissions.administrator


async def is_moderator(ctx: commands.Context):
    guild_roles = []
    for role in ctx.guild.roles:
        guild_roles.append(role.name)

    moderator_roles = guild_roles[guild_roles.index('Moderator'):]
    return ctx.author.guild_permissions.administrator or (ctx.author.top_role in moderator_roles)


class RoleRequestor(commands.Cog):
    def __init__(self, bot: commands.Bot):
        self.bot = bot

    async def build_category_message(self, ctx: commands.Context, category: tuple):
        roles = self.bot.dbc.execute("SELECT role, description, reaction FROM roles WHERE categoryId=?",
                                     (category[0],)).fetchall()
        message_lines = [
            f"\n**{category[2]} Roles:**"
        ]

        if len(roles) == 0:
            message_lines.append('\n\nThere are no roles in this category.')
            return ''.join(message_lines)

        for db_role in roles:
            role = discord.utils.get(ctx.guild.roles, name=db_role[0])
            if not role:
                await ctx.send(f"@FarrakKilhn Something weird happened. Role {db_role[0]} should exist, but doesn't.")
                return

            # Append role line to message body
            emoji = db_role[2] if len(db_role[2]) == 1 else f"<{db_role[2]}>"
            description = f": *{db_role[1]}*" if db_role[1] else None
            message_lines.append(f"\n\n{emoji} - {role.mention}{description}")

        message_lines.append("\n\n")
        return ''.join(message_lines)

    async def fetch_role(self, event: discord.RawReactionActionEvent, guild: discord.Guild, member: discord.Member):
        db_role = self.bot.dbc.execute("SELECT role FROM roles r "
                                       "JOIN role_categories rc ON r.categoryId=rc.id "
                                       "WHERE rc.guildId=? "
                                       "   AND rc.messageId=? "
                                       "   AND r.reaction=?",
                                       (event.guild_id, event.message_id, event.emoji.name)).fetchone()
        if not db_role:
            # Alert user that role cannot be added
            await member.send("The role you requested could not be added, sorry.")
            raise Exception(f"Role name could not be found in database based on guild and "
                            f"message id for guild {guild.name} with id {guild.id}")

        role = discord.utils.get(guild.roles, name=db_role[0])
        if not role:
            # Alert user that role cannot be added
            await member.send("The role you requested could not be added, sorry.")
            raise Exception(f"Discord role could not be resolved from role name in guild "
                            f"{guild.name} with id {guild.id}")

        return role

    @commands.command(
        name="init-role-system",
        brief="Create a #role-request channel for users to interact with AginahBot and request roles",
        help="Create a #role-request text channel for users to interact with AginahBot and request roles. "
             "This channel will be used to post role category messages users can react to to add or remove roles\n"
             "Usage: !aginah init-role-system")
    @commands.check(is_administrator)
    async def init_role_system(self, ctx: commands.Context):
        mod_role = discord.utils.get(ctx.guild.roles, name=MODERATOR_ROLE)
        if not mod_role:
            await ctx.send("There is no @Moderator role on your server. Please create one and try again.")
            return

        for channel in ctx.guild.text_channels:
            if channel.name == ROLE_REQUEST_CHANNEL:
                await ctx.send("It looks like the roles system is already active on this server.")
                return

        text_channel = await ctx.guild.create_text_channel(ROLE_REQUEST_CHANNEL,
                                                           topic="Assign yourself roles if you would like to be pinged",
                                                           overwrites={
                                                               ctx.guild.default_role: discord.PermissionOverwrite(
                                                                   add_reactions=False),
                                                               ctx.guild.me: discord.PermissionOverwrite(
                                                                   add_reactions=True)
                                                           })
        await text_channel.send(f"The following roles are available on this server. If you would like to be "
                                f"assigned a role, please react to this message with the indicated emoji. All "
                                f"roles are pingable by everyone on the server. Remember, with great power comes "
                                f"great responsibility.\n")
        await ctx.send("Role system initialized.")

    @commands.command(
        name="destroy-role-system",
        brief="Delete the role-request channel and all categories and permissions created by this bot",
        help="Delete the role-request channel and all categories and permissions created by this bot\n"
             "Usage: !aginah init-role-system")
    @commands.check(is_administrator)
    async def destroy_role_system(self, ctx: commands.Context):
        # Fetch categories for this guild
        categories = self.bot.dbc.execute("SELECT id FROM role_categories WHERE guildId=?", (ctx.guild.id,))
        for category in categories:
            # Fetch roles for this category
            db_roles = self.bot.dbc.execute("SELECT role FROM roles WHERE categoryId=?", (category[0],))
            for db_role in db_roles:
                # Delete role from the guild
                role = discord.utils.get(ctx.guild.roles, name=db_role[0])
                await role.delete()

            # Delete roles from the db
            self.bot.dbc.execute("DELETE FROM roles WHERE categoryId=?", (category[0],))

        # Delete categories from db
        self.bot.dbc.execute("DELETE FROM role_categories WHERE guildId=?", (ctx.guild.id,))
        self.bot.db.commit()

        channel = discord.utils.get(ctx.guild.text_channels, name=ROLE_REQUEST_CHANNEL)
        await channel.delete()

        await ctx.send("Role system destroyed.")

    @commands.command(
        name="create-role-category",
        brief="Create a category for roles to be added to.",
        help="Create a category for roles to be added to. Each category will have its own message in the "
             "#role-request channel. Category names must be a single alphanumeric word.\n"
             "Usage: !aginah create-role-category CategoryName"
    )
    @commands.check(is_moderator)
    async def create_role_category(self, ctx: commands.Context):
        args = ctx.message.content.split(' ')
        if len(args) != 3:
            await ctx.send("Looks like you have a syntax error! `!aginah help create-role-category` for more info.")
            return

        # If category already exists, warn user and do nothing
        if self.bot.dbc.execute("SELECT 1 FROM role_categories WHERE guildId=? AND category=?",
                                (ctx.guild.id, args[2],)).fetchone():
            await ctx.send("That category already exists!")
            return

        # Create the new category and add a message to the #role-request channel
        role_channel = discord.utils.get(ctx.guild.text_channels, name=ROLE_REQUEST_CHANNEL)
        message = await role_channel.send(f"Creating category. Standby...")

        # Save the new category and message id into the database
        self.bot.dbc.execute("INSERT INTO role_categories (guildId, category, messageId) VALUES (?,?,?)",
                             (ctx.guild.id, args[2].strip(), message.id))
        self.bot.db.commit()

        category = self.bot.dbc.execute("SELECT * FROM role_categories WHERE guildId=? AND category=?",
                                        (ctx.guild.id, args[2])).fetchone()

        await message.edit(content=await self.build_category_message(ctx, category))

        await ctx.send("Category created.")

    @commands.command(
        name="delete-role-category",
        brief="Delete a role category.",
        help="Delete a role category. All roles within this category will also be deleted.\n"
             "Usage: !aginah delete-role-category CategoryName"
    )
    @commands.check(is_moderator)
    async def delete_role_category(self, ctx: commands.Context):
        args = ctx.message.content.split(' ')
        if len(args) != 3:
            await ctx.send("Looks like you have a syntax error! `!aginah help create-role-category` for more info.")
            return

        # Fetch requisite data (role channel, category db row, role db rows)
        role_channel = discord.utils.get(ctx.guild.text_channels, name=ROLE_REQUEST_CHANNEL)
        category = self.bot.dbc.execute("SELECT id, messageId FROM role_categories WHERE guildId=? AND category=?",
                                        (ctx.guild.id, args[2])).fetchone()

        # If category does not exist, warn user and do nothing
        if not category:
            await ctx.send("That category does not exist!")
            return

        roles = self.bot.dbc.execute("SELECT * FROM roles WHERE categoryId=?", (category[0],))

        # Delete roles from Discord
        for role in roles:
            await (discord.utils.get(ctx.guild.roles, name=role.role)).delete()

        # Delete roles from db
        self.bot.dbc.execute("DELETE FROM roles WHERE categoryId=?", (category[0],))
        self.bot.db.commit()

        # Delete category message
        message = await role_channel.fetch_message(category[1])
        await message.delete()

        # Delete category
        self.bot.dbc.execute("DELETE FROM role_categories WHERE id=?", (category[0],))
        self.bot.db.commit()

        await ctx.send("Category deleted.")

    @commands.command(
        name="create-role",
        brief="Create a role which may be pinged",
        help="Create a role which may be pinged by anyone on the server\n"
             "Usage: !aginah create-role CategoryName RoleName reaction [description]"
    )
    @commands.check(is_moderator)
    async def create_role(self, ctx: commands.Context):
        args = ctx.message.content.split(' ')
        if len(args) < 5:
            await ctx.send("Looks like you have a syntax error! `!aginah help create-role` for more info.")
            return

        # Ensure role does not exist
        for role in ctx.guild.roles:
            if role.name == args[3]:
                await ctx.send("That role already exists!")
                return

        emoji = args[4] if len(args[4]) == 1 else args[4][1:-1]

        # Fetch category
        category = self.bot.dbc.execute("SELECT * FROM role_categories WHERE guildId=? AND category=?",
                                        (ctx.guild.id, args[2])).fetchone()
        if not category:
            await ctx.send("That category doesn't exist!")
            return

        # Fetch role request channel
        channel = discord.utils.get(ctx.guild.text_channels, name=ROLE_REQUEST_CHANNEL)
        if not channel:
            await ctx.send("It looks like the role requestor is not set up on this server. Please contact an admin.")
            return

        # Fetch role message
        message = await channel.fetch_message(category[3])
        if not message:
            await ctx.send("Something is very wrong here. Hey @FarrakKilhn take a look at this.")
            return

        try:
            # Add reaction to category message
            await message.add_reaction(emoji)
        except:
            await ctx.send("The emoji you tried to use is not available on this server. Please try again.")
            return

        # Add role to Discord
        role = await ctx.guild.create_role(name=args[3], mentionable=True)

        # Add db row
        desc = ' '.join(args[5:]) if len(args) > 5 else None
        self.bot.dbc.execute("INSERT INTO roles (categoryId, role, reaction, description) VALUES (?,?,?,?)",
                             (category[0], role.name.strip(), emoji, desc.strip()))
        self.bot.db.commit()

        # Update message contents to show updated roles
        await message.edit(content=await self.build_category_message(ctx, category))

        await ctx.send("Role created.")

    @commands.command(
        name="delete-role",
        brief="Delete a role created by this bot",
        help="Delete a role created by this bot\n"
             "Usage: !aginah delete-role CategoryName RoleName"
    )
    @commands.check(is_moderator)
    async def delete_role(self, ctx: commands.Context):
        args = ctx.message.content.split(' ')
        if len(args) != 4:
            await ctx.send("Looks like you have a syntax error! `!aginah help delete-role` for more info.")
            return

        role = discord.utils.get(ctx.guild.roles, name=args[3])
        if not role:
            await ctx.send("That role does not exist!")
            return

        # Fetch category
        category = self.bot.dbc.execute("SELECT * FROM role_categories WHERE guildId=? AND category=?",
                                        (ctx.guild.id, args[2],)).fetchone()
        if not category:
            await ctx.send("That category doesn't exist!")
            return

        # Fetch role request channel
        channel = discord.utils.get(ctx.guild.text_channels, name=ROLE_REQUEST_CHANNEL)
        if not channel:
            await ctx.send(
                "It looks like the role requestor is not set up on this server. Please contact an admin.")
            return

        # Fetch role message
        message = await channel.fetch_message(category[3])
        if not message:
            await ctx.send("Something is very wrong here. Hey @FarrakKilhn take a look at this.")
            return

        # Fetch db role
        db_role = self.bot.dbc.execute("SELECT * FROM roles WHERE categoryId=? and role=?",
                                       (category[0], args[3])).fetchone()

        # Remove role from db
        self.bot.dbc.execute("DELETE FROM roles WHERE categoryId=? AND role=?", (category[0], role.name))
        self.bot.db.commit()

        # Delete Discord role
        await role.delete()

        # Edit category message to display updated roles
        await message.edit(content=await self.build_category_message(ctx, category))

        # Remove role reactions from message
        for reaction in message.reactions:
            if reaction.emoji == db_role[3]:
                await reaction.clear()

        # Notify of success
        await ctx.send("Role deleted.")

    @commands.Cog.listener()
    async def on_raw_reaction_add(self, event: discord.RawReactionActionEvent):
        # Do not operate on messages which are not a category message
        skip = True
        valid_ids = self.bot.dbc.execute("SELECT messageId FROM role_categories")
        for msgId in valid_ids:
            if msgId[0] == event.message_id:
                skip = False
                break
        if skip:
            return

        # Fetch guild and member info
        guild = discord.utils.get(self.bot.guilds, id=event.guild_id)
        member = discord.utils.get(guild.members, id=event.user_id)

        # If a bot added a reaction (probably this bot), do nothing
        if member.bot:
            return

        if not guild or not member:
            raise Exception("Unable to determine guild or member to add role.")

        role = await self.fetch_role(event, guild, member)
        await member.add_roles(role)

    @commands.Cog.listener()
    async def on_raw_reaction_remove(self, event: discord.RawReactionActionEvent):
        # Do not operate on messages which are not a category message
        skip = True
        valid_ids = self.bot.dbc.execute("SELECT messageId FROM role_categories")
        for msgId in valid_ids:
            if msgId[0] == event.message_id:
                skip = False
                break
        if skip:
            return

        # Fetch guild and member info
        guild = discord.utils.get(self.bot.guilds, id=event.guild_id)
        member = discord.utils.get(guild.members, id=event.user_id)

        # If a bot added a reaction (probably this bot), do nothing
        if member.bot:
            return

        if not guild or not member:
            raise Exception("Unable to determine guild or member to add role.")

        role = await self.fetch_role(event, guild, member)
        await member.remove_roles(role)


# All cogs must have this function
def setup(bot: commands.Bot):
    bot.add_cog(RoleRequestor(bot))

File: commandCategories/roleRequestor.js
const config = require('../config.json');
const {generalErrorHandler} = require('../errorHandlers');
const { parseEmoji, dbQueryOne, dbQueryAll, dbExecute } = require('../lib');

const updateCategoryMessage = async (client, guild, messageId) => {
  // Fetch the target message
  let sql = `SELECT rc.id, rc.categoryName, rs.roleRequestChannelId FROM role_categories rc
             JOIN role_systems rs ON rc.roleSystemId = rs.id
             JOIN guild_data gd ON rs.guildDataId = gd.id
             WHERE gd.guildId=?
               AND rc.messageId=?`;
  const roleCategory = await dbQueryOne(sql, [guild.id, messageId]);
  if (!roleCategory) { throw Error("Unable to update category message. Role category could not be found."); }

  const roleInfo = [`> __${roleCategory.categoryName}__`];
  sql = `SELECT r.roleId, r.reaction, r.description FROM roles r WHERE r.categoryId=?`;
  const roles = await dbQueryAll(sql, [roleCategory.id]);
  roles.forEach((role) => {
    roleInfo.push(`> ${role.reaction} ${guild.roles.resolve(role.roleId)}` +
        `${role.description ? `: ${role.description}` : ''}`);
  });

  // If there are no roles in this category, mention that there are none
  if (roleInfo.length === 1) { roleInfo.push("> There are no roles in this category yet."); }

  // Fetch and edit the category message
  guild.channels.resolve(roleCategory.roleRequestChannelId).messages.fetch(messageId)
      .then((categoryMessage) => categoryMessage.edit(roleInfo)
          .then().catch((err) => generalErrorHandler(err)));
};

module.exports = {
  category: 'Role Requestor',
  commands: [
    {
      name: 'init-role-system',
      description: 'Create a role-request channel for users to interact with AginahBot and request roles.',
      longDescription: 'Create a #role-request text channel for users to interact with AginahBot and request ' +
        'roles. This channel will be used to post role category messages users can react to to add or ' +
        'remove roles.',
      aliases: ['irs'],
      usage: '`!aginah init-role-system`',
      minimumRole: null,
      adminOnly: true,
      guildOnly: true,
      async execute(message) {
        let sql = `SELECT gd.id AS guildDataId, rs.id AS roleSystemId
                    FROM guild_data gd
                    LEFT JOIN role_systems rs ON gd.id = rs.guildDataId
                    WHERE gd.guildId=?`;
        const row = await dbQueryOne(sql, [message.guild.id]);
        if (!row) { throw new Error(); }
        if (row.roleSystemId) {
          return message.channel.send("The role system has already been set up on your server.");
        }
        message.guild.channels.create('role-request', {
          type: 'text',
          topic: 'Request roles so that you may be pinged for various notifications.',
          reason: 'Role Request system created.',
          permissionOverwrites: [
            {
              id: message.client.user.id,
              allow: [ 'SEND_MESSAGES', 'ADD_REACTIONS' ],
            },
            {
              id: message.guild.roles.everyone.id,
              deny: [ 'SEND_MESSAGES', 'ADD_REACTIONS' ]
            }
          ],
        }).then(async (channel) => {
          // Add role system data to the database
          let sql = `INSERT INTO role_systems (guildDataId, roleRequestChannelId) VALUES(?, ?)`
          await dbExecute(sql, [row.guildDataId, channel.id]);
          channel.send(`The following roles are available on this server. If you would like to be ` +
              `assigned a role, please react to the appropriate message with the indicated ` +
              `emoji. All roles are pingable by everyone on the server. Remember, with great ` +
              `power comes great responsibility.`);
        }).catch((error) => generalErrorHandler(error));
      }
    },
    {
      name: 'destroy-role-system',
      description: 'Delete the role-request channel and all categories and permissions created by this bot.',
      longDescription: null,
      aliases: ['drs'],
      usage: '`!aginah destroy-role-system`',
      minimumRole: null,
      adminOnly: true,
      guildOnly: true,
      async execute(message) {
        let sql = `SELECT rs.id, rs.roleRequestChannelId FROM role_systems rs
                    JOIN guild_data gd ON rs.guildDataId = gd.id
                    WHERE gd.guildId=?`;
        const roleSystem = await dbQueryOne(sql, [message.guild.id]);
        if (!roleSystem) {
          // If the role system does not exist, there is no need to attempt to delete it
          return message.channel.send("The role system is not currently installed on this server.");
        }

        // Loop over the role categories and delete them
        sql = `SELECT id FROM role_categories WHERE roleSystemId=?`;
        const roleCategories = await dbQueryAll(sql, [roleSystem.id]);
        for (let roleCategory of roleCategories) {
          // Loop over the roles in each category and delete them from the server
          const roles = await dbQueryAll(`SELECT roleId FROM roles WHERE categoryId=?`, [roleCategory.id]);
          for (let role of roles) {
            await message.guild.roles.resolve(role.roleId).delete();
          }
          await dbExecute(`DELETE FROM roles WHERE categoryId=?`, [roleCategory.id]);
        }

        // Database cleanup
        await dbExecute(`DELETE FROM role_categories WHERE roleSystemId=?`, [roleSystem.id]);
        message.guild.channels.resolve(roleSystem.roleRequestChannelId).delete();
        await dbExecute(`DELETE FROM role_systems WHERE id=?`, [roleSystem.id]);
      }
    },
    {
      name: 'create-role-category',
      description: 'Create a category for roles to be added to.',
      longDescription: `Create a category for roles to be added to. Each category will have its own message ` +
        `in the #role-request channel. Category names must be a single alphanumeric word.`,
      aliases: [],
      usage: '`!aginah create-role-category CategoryName`',
      minimumRole: config.moderatorRole,
      adminOnly: false,
      guildOnly: true,
      async execute(message, args) {
        if (args.length === 0) {
          return message.channel.send("You must specify a name for the category!");
        }

        let sql = `SELECT 1 FROM role_categories rc
                    JOIN role_systems rs ON rc.roleSystemId = rs.id
                    JOIN guild_data gd ON rs.guildDataId = gd.id
                    WHERE gd.guildId=?
                      AND rc.categoryName=?`;
        const row = await dbQueryOne(sql, [message.guild.id, args[0]]);
        if (row) { return message.channel.send("That category already exists!"); }

        sql = `SELECT rs.id, rs.roleRequestChannelId
                      FROM role_systems rs
                      JOIN guild_data gd ON rs.guildDataId=gd.id
                      WHERE gd.guildId=?`;
        const roleSystem = await dbQueryOne(sql, [message.guild.id]);
        if (!roleSystem) {
          return message.channel.send("The role system has not been setup on this server.");
        }

        // Add the category message to the #role-request channel
        message.guild.channels.resolve(roleSystem.roleRequestChannelId).send("Creating new category...")
          .then(async (categoryMessage) => {
            // Update database with new category message data
            let sql = `INSERT INTO role_categories (roleSystemId, categoryName, messageId) VALUES (?, ?, ?)`;
            await dbExecute(sql, [roleSystem.id, args[0], categoryMessage.id]);
            await updateCategoryMessage(message.client, message.guild, categoryMessage.id);
          });
      }
    },
    {
      name: 'delete-role-category',
      description: 'Delete a role category.',
      longDescription: 'Delete a role category. All roles within this category will also be deleted.',
      aliases: [],
      usage: '`!aginah delete-role-category CategoryName`',
      minimumRole: config.moderatorRole,
      adminOnly: false,
      guildOnly: true,
      async execute(message, args) {
        if (args.length === 0) {
          return message.channel.send("You must specify the category to delete!");
        }

        // Get the role category data
        let sql = `SELECT rc.id, rc.messageId, rs.roleRequestChannelId FROM role_categories rc
                    JOIN role_systems rs ON rc.roleSystemId = rs.id
                    JOIN guild_data gd ON rs.guildDataId = gd.id
                    WHERE gd.guildId=?
                      AND rc.categoryName=?`;
        const roleCategory = await dbQueryOne(sql, [message.guild.id, args[0]]);
        if (!roleCategory) { return message.channel.send("That category does not exist!"); }
        await dbExecute(`DELETE FROM roles WHERE categoryId=?`, [roleCategory.id]);
        await dbExecute(`DELETE FROM role_categories WHERE id=?`, [roleCategory.id]);
        message.guild.channels.resolve(roleCategory.roleRequestChannelId).messages.fetch(roleCategory.messageId)
          .then((categoryMessage) => categoryMessage.delete()).catch((err) => generalErrorHandler(err));
      }
    },
    {
      name: 'create-role',
      description: 'Create a pingable role.',
      longDescription: null,
      aliases: [],
      usage: '`!aginah cmd create-role CategoryName RoleName Reaction [Description]`',
      minimumRole: config.moderatorRole,
      adminOnly: false,
      guildOnly: true,
      async execute(message, args) {
        // Check for existing role
        let sql = `SELECT 1 FROM roles r
                    JOIN role_categories rc ON r.categoryId = rc.id
                    JOIN role_systems rs ON rc.roleSystemId = rs.id
                    JOIN guild_data gd ON rs.guildDataId = gd.id
                    WHERE gd.guildId=?
                      AND rc.categoryName=?
                      AND r.roleName=?`;
        const row = await dbQueryOne(sql, [message.guild.id, args[0], args[1]]);
        // If the role already exists, do nothing
        if (row) { return message.channel.send("That role already exists!"); }

        // Verify the requested emoji is available on this server
        const emoji = parseEmoji(message.guild, args[2]);
        if (!emoji) { return message.channel.send("That emoji is not available on this server!"); }

        sql = `SELECT rc.id, rc.messageId, rs.roleRequestChannelId FROM role_categories rc
                      JOIN role_systems rs ON rc.roleSystemId = rs.id
                      JOIN guild_data gd ON rs.guildDataId = gd.id
                      WHERE gd.guildId=?
                        AND rc.categoryName=?`;
        const roleCategory = await dbQueryOne(sql, [message.guild.id, args[0]]);
        // If there is no such category, warn the user and do nothing
        if (!roleCategory) { return message.channel.send("That category doesn't exist!"); }

        // Create the role on the server
        message.guild.roles.create({
          data: {
            name: args[1],
            mentionable: true,
          },
          reason: 'Added as part of role-request system.',
        }).then(async (role) => {
          // Add the role to the database
          await dbExecute(
            `INSERT INTO roles (categoryId, roleId, roleName, reaction, description) VALUES (?, ?, ?, ?, ?)`,
            [roleCategory.id, role.id, args[1], emoji.toString(), (args[3] ? args.slice(3).join(' ') : null)]
          );
          await updateCategoryMessage(message.client, message.guild, roleCategory.messageId);

          // Add the reaction to the category message
          message.guild.channels.resolve(roleCategory.roleRequestChannelId).messages.fetch(roleCategory.messageId)
              .then((categoryMessage) => categoryMessage.react(emoji).catch((err) => generalErrorHandler(err)));
        }).catch((error) => {
          throw new Error(`Unable to create guild role. ${error}`);
        });
      }
    },
    {
      name: 'modify-role-reaction',
      description: 'Alter the reaction associated with a role created by this bot.',
      longDescription: null,
      aliases: [],
      usage: '`!aginah modify-role-reaction CategoryName RoleName Reaction`',
      minimumRole: config.moderatorRole,
      adminOnly: false,
      guildOnly: true,
      async execute(message, args) {
        // Check for existing role
        let sql = `SELECT r.id, rc.messageId, rs.roleRequestChannelId, r.reaction FROM roles r
                    JOIN role_categories rc ON r.categoryId = rc.id
                    JOIN role_systems rs ON rc.roleSystemId = rs.id
                    JOIN guild_data gd ON rs.guildDataId = gd.id
                    WHERE gd.guildId=?
                      AND rc.categoryName=?
                      AND r.roleName=?`;
        const role = await dbQueryOne(sql, [message.guild.id, args[0], args[1]]);
        // If the role already exists, do nothing
        if (!role) { return message.channel.send("That role does not exist!"); }

        // Verify the requested emoji is available on this server
        const emoji = parseEmoji(message.guild, args[2]);
        if (!emoji) { return message.channel.send("That emoji is not available on this server!"); }

        // Remove reactions from the role category message
        message.guild.channels.resolve(role.roleRequestChannelId).messages.fetch(role.messageId)
          .then((categoryMessage) => {
            categoryMessage.reactions.cache.each((reaction) => {
              if (reaction.emoji.toString() === role.reaction) { reaction.remove(); }
            });

            // Add new reaction to message
            categoryMessage.react(emoji);
          }).catch((err) => generalErrorHandler(err));

        await dbExecute(`UPDATE roles SET reaction=? WHERE id=?`, [emoji.toString(), role.id]);
        await updateCategoryMessage(message.client, message.guild, role.messageId);
      }
    },
    {
      name: 'modify-role-description',
      description: 'Alter the description associated with a role created by this bot.',
      longDescription: null,
      aliases: [],
      usage: '`!aginah modify-role-description CategoryName RoleName [Description]`',
      minimumRole: config.moderatorRole,
      adminOnly: false,
      guildOnly: true,
      async execute(message, args) {
        // Check for existing role
        let sql = `SELECT r.id, rc.messageId, rs.roleRequestChannelId FROM roles r
                    JOIN role_categories rc ON r.categoryId = rc.id
                    JOIN role_systems rs ON rc.roleSystemId = rs.id
                    JOIN guild_data gd ON rs.guildDataId = gd.id
                    WHERE gd.guildId=?
                      AND rc.categoryName=?
                      AND r.roleName=?`;
        const role = await dbQueryOne(sql, [message.guild.id, args[0], args[1]]);
        if (!role) { return message.channel.send("That role does not exist!"); }

        sql = `UPDATE roles SET description=? WHERE id=?`;
        await dbExecute(sql, [(args[2] ? args.slice(2).join(' ') : null), role.id]);
        await updateCategoryMessage(message.client, message.guild, role.messageId);
      }
    },
    {
      name: 'delete-role',
      description: 'Delete a role created by this bot.',
      longDescription: null,
      aliases: [],
      usage: '`!aginah delete-role CategoryName RoleName`',
      minimumRole: config.moderatorRole,
      adminOnly: false,
      guildOnly: true,
      async execute(message, args) {
        let sql = `SELECT r.id, rc.id AS categoryId, rc.messageId, r.reaction, r.roleId, rs.roleRequestChannelId
                    FROM roles r
                    JOIN role_categories rc ON r.categoryId = rc.id
                    JOIN role_systems rs ON rc.roleSystemId = rs.id
                    JOIN guild_data gd ON rs.guildDataId = gd.id
                    WHERE gd.guildId=?
                      AND rc.categoryName=?
                      AND r.roleName=?`;
        const role = await dbQueryOne(sql, [message.guild.id, args[0], args[1]]);
        if (!role) { return message.channel.send("That role does not exist!"); }

        // Remove reactions from the role category message
        message.guild.channels.resolve(role.roleRequestChannelId).messages.fetch(role.messageId)
          .then((categoryMessage) => categoryMessage.reactions.cache.array().forEach((r) => {
            if (r.emoji.toString() === role.reaction) { r.remove(); }
          }))
          .catch((err) => generalErrorHandler(err));

        // Remove the role from the guild
        message.guild.roles.resolve(role.roleId).delete();

        // Delete rows from the roles table and update role category message
        await dbExecute(`DELETE FROM roles WHERE id=? AND categoryId=?`, [role.id, role.categoryId]);
        await updateCategoryMessage(message.client, message.guild, role.messageId);
      }
    },
  ],
};
